File: frontend/src/api/agent-performance.ts
import api from './index'

// Workspace-wide aggregated metrics.
export interface AgentPerformanceOverview {
  workspace_id: number
  total_tasks: number
  completed_tasks: number
  failed_tasks: number
  cancelled_tasks: number
  running_tasks: number
  pending_tasks: number
  success_rate: number          // 0-100
  avg_duration_seconds: number
  total_duration_seconds: number
  period_start?: string
  period_end?: string
}

// Per-template performance breakdown.
export interface TemplatePerformance {
  agent_template_id?: number
  template_name: string
  task_type?: string
  total_tasks: number
  completed_tasks: number
  failed_tasks: number
  cancelled_tasks: number
  success_rate: number
  avg_duration_seconds: number
  last_run_at?: string
}

// Time-bucketed metric point for trend charts.
export interface TimelinePoint {
  bucket_start: string
  bucket_end: string
  total_tasks: number
  completed_tasks: number
  failed_tasks: number
  success_rate: number
  avg_duration_seconds: number
}

// Failure reason tally.
export interface FailureBreakdown {
  failure_reason: string
  count: number
  percentage: number
}

export type BucketGranularity = 'day' | 'week' | 'month'

export interface PeriodFilter {
  from?: string  // RFC3339
  to?: string    // RFC3339
}

function toParams(period: PeriodFilter, extra: Record<string, string> = {}): Record<string, string> {
  const params: Record<string, string> = { ...extra }
  if (period.from) params.from = period.from
  if (period.to) params.to = period.to
  return params
}

export const agentPerformanceApi = {
  overview(workspaceId: number, period: PeriodFilter = {}): Promise<AgentPerformanceOverview> {
    return api
      .get(`/workspaces/${workspaceId}/agent-performance/overview`, { params: toParams(period) })
      .then(res => res.data)
  },

  byTemplate(workspaceId: number, period: PeriodFilter = {}): Promise<TemplatePerformance[]> {
    return api
      .get(`/workspaces/${workspaceId}/agent-performance/by-template`, { params: toParams(period) })
      .then(res => res.data)
  },

  timeline(
    workspaceId: number,
    bucket: BucketGranularity = 'day',
    period: PeriodFilter = {}
  ): Promise<TimelinePoint[]> {
    return api
      .get(`/workspaces/${workspaceId}/agent-performance/timeline`, {
        params: toParams(period, { bucket })
      })
      .then(res => res.data)
  },

  failureBreakdown(workspaceId: number, period: PeriodFilter = {}): Promise<FailureBreakdown[]> {
    return api
      .get(`/workspaces/${workspaceId}/agent-performance/failures`, { params: toParams(period) })
      .then(res => res.data)
  }
}

File: backend/internal/dto/response/agent_performance.go
package response

import "time"

// AgentPerformanceOverview aggregates workspace-wide agent task execution metrics.
type AgentPerformanceOverview struct {
	WorkspaceID        uint64  `json:"workspace_id"`
	TotalTasks         int64   `json:"total_tasks"`
	CompletedTasks     int64   `json:"completed_tasks"`
	FailedTasks        int64   `json:"failed_tasks"`
	CancelledTasks     int64   `json:"cancelled_tasks"`
	RunningTasks       int64   `json:"running_tasks"`
	PendingTasks       int64   `json:"pending_tasks"`
	SuccessRate        float64 `json:"success_rate"`         // 0-100, completed/(completed+failed)
	AvgDurationSeconds float64 `json:"avg_duration_seconds"` // mean of completed tasks
	TotalDurationSec   float64 `json:"total_duration_seconds"`
	PeriodStart        *time.Time `json:"period_start,omitempty"`
	PeriodEnd          *time.Time `json:"period_end,omitempty"`
}

// TemplatePerformance groups metrics per agent template.
type TemplatePerformance struct {
	AgentTemplateID    *uint64 `json:"agent_template_id,omitempty"`
	TemplateName       string  `json:"template_name"`
	TaskType           string  `json:"task_type,omitempty"`
	TotalTasks         int64   `json:"total_tasks"`
	CompletedTasks     int64   `json:"completed_tasks"`
	FailedTasks        int64   `json:"failed_tasks"`
	CancelledTasks     int64   `json:"cancelled_tasks"`
	SuccessRate        float64 `json:"success_rate"`
	AvgDurationSeconds float64 `json:"avg_duration_seconds"`
	LastRunAt          *time.Time `json:"last_run_at,omitempty"`
}

// TimelinePoint represents an aggregated metrics bucket over a time range.
type TimelinePoint struct {
	BucketStart         time.Time `json:"bucket_start"`
	BucketEnd           time.Time `json:"bucket_end"`
	TotalTasks          int64     `json:"total_tasks"`
	CompletedTasks      int64     `json:"completed_tasks"`
	FailedTasks         int64     `json:"failed_tasks"`
	SuccessRate         float64   `json:"success_rate"`
	AvgDurationSeconds  float64   `json:"avg_duration_seconds"`
}

// FailureBreakdown tallies failure reasons for diagnostic dashboards.
type FailureBreakdown struct {
	FailureReason string `json:"failure_reason"`
	Count         int64  `json:"count"`
	Percentage    float64 `json:"percentage"` // share of total failures
}
